File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.cc
#define _CRT_SECURE_NO_WARNINGS
#include <stdio.h>
#include <stdlib.h>

int main(void) {
	int i;
	int N, K;
	int count, ans;
	long long* arr;
	long long max = 0;
	long long left, right, mid;

	scanf("%d %d", &N, &K);

	arr = (long long*)malloc(sizeof(long long) * N);

	for (i = 0; i < N; i++) {
		scanf("%lld", &arr[i]);
		if (arr[i] > max) max = arr[i];
	}

	ans = 0;
	left = 1;
	right = max;

	while (left <= right) {
		count = 0;
		mid = (left + right) / 2;

		for (i = 0; i < N; i++) {
			count += arr[i] / mid;
		}

		if (count >= K) {
			left = mid + 1;
			if (ans < mid) ans = mid;
		}
		else right = mid - 1;
	}
	printf("%d\n", ans);

	free(arr);
	return 0;
}

File: 백준/Silver/1654. 랜선 자르기/랜선 자르기.java
import java.io.*;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringTokenizer st = new StringTokenizer(br.readLine());

        // 입력
        int K = Integer.parseInt(st.nextToken());
        int N = Integer.parseInt(st.nextToken());
        long[] arr = new long[K + 1];
        long right = 0;
        for (int i = 0; i < K; i++) {
            arr[i] = Integer.parseInt(br.readLine());
            right = Math.max(arr[i], right);
        }
        //
        long left = 1;
        long result = 0;
        while (left <= right) {
            long count = 0;
            long mid = (left + right) / 2;
            for (int i = 0; i < K; i++) {
                count += arr[i] / mid;
            }
            if (count < N) {
                // 길이 줄이기
                right = mid - 1;
            }
            else {
                // 길이 늘리기
                left = mid + 1;
                if (result < mid) result = mid;
            }
        }
        // 출력
        System.out.println(result);
    }
}